accepted/PSR-7-http-message-meta.md

```diff
@@ -691,3 +691,14 @@ Further characters or sequences in header values should be rejected according
 to the HTTP specification.
 
 [1]: https://datatracker.ietf.org/doc/html/rfc7230#section-3.2
+
+### 7.2 urlencode User Info
+
+In any release, `UriInterface::withUserInfo` MUST NOT double urlencode the username
+and password if they are already urlencoded.
+
+In release 1.1, `UriInterface::withUserInfo` SHOULD urlencode username or password when
+they contain characters that are invalid in an URL and are not already urlencoded.
```

Comment:
to sum up the above discussion: would it be correct to change this to something like: `SHOULD escape characters in in username and password that are defined as reserved in [RFC3986](https://www.rfc-editor.org/rfc/rfc3986#appendix-A)`

and accordingly below, and also in the rule about double-escaping?

Reply:
We might also say SHOULD NOT encode other characters, that would cover double encoding and a bit more, eg `\` which is encoded by guzzle (but not by nyholm's)

Reply:
With encoding only reserved characters, if you pass `foo%3Abar:baz` as username, it will be encoded to `foo%3Abar%3Abaz`. How would you revert it back to `foo%3Abar:baz`?

Reply:
@rob006 then the correct way to formulate would be: if it contains any reserved character, the whole value should be encoded? if you pass `foo%eAbar:baz`, the implementation has to assume that the string is not yet encoded, and should therefore also encode the `%` i think.

damn, i don't think there really is a good solution... because if we do that, the behaviour is different if the caller never encodes based on whether they pass `foo%3Abar` or `foo%3Abar:baz` (because the receiving server will urldecode and translate the `%3A`.

nonetheless, i think we should figure out the rule that makes for the least confusion and add an errata that clarifies how to do this. and in the interest of internal consistence and low surprises, i would go with the rules outlined on `getQuery`. (i am not sure what those rules mean for the `foo%3Abar:baz` example. but the current implementations i think will do a compromise that will not work, as they only encode the `:` but not the `%`)

Reply:
There is simple and reliable solution: always encode. You build user info by encoding username and password, and implode them using `:`. If you want to retrieve username and password, you do the same backwards: explode results of `getUserInfo()` by `:`, and decode each part. Yes, it is inconvenient if you want to pass `getUserInfo()` result to `withUserInfo()`, but this is a result of asymmetry between these two methods, and can't be really fixed without serious BC break. But this use case could be simplified by adding `getUserName()` and `getPassword()` to interface: `$uri->withUserInfo($previous->getUserName(), $previous->getPassword()))`.

Reply:
by `you`, do you mean the caller or the psr7 implementation?

given that all the largely used implementations discussed above do encode conditionally, i think clarifying that the arguments always must be encoded by the caller is not a viable option as it conflicts with existing implementations. if implementations were to change that behaviour, this would be a nasty hidden BC break for users.

my aim with this pull request is to try to clarify what implementations should do in a way that creates as little friction as possible.

i do think there would be merit for a new PSR with stricter and more robust rules, but that is out of scope for this pull request and discussion. (please feel free to cite this discussion as one motivation why there should be a new PSR for Uri, should you propose such a PSR)

Reply:
> by `you`, do you mean the caller or the psr7 implementation?

I mean implementation. So `withUserInfo()` should always encode and `getUserInfo()` should always return this encoded value.  `getUserName()` and `getPassword()` would return **decoded** values (in the same form as passed to `withUserInfo()`).

Reply:
that would be consistent and deterministic. we can't add new methods afaik, though if it is allowed for a version 2.0 it would make sense.

but i am unsure if that is what the current implementations do. the phpdoc on UriInterface::getQuery says `The value returned MUST be percent-encoded, but MUST NOT double-encode any characters.` i am not 100% sure i understand correctly, but i think when this was written, the `foo%3Abar:baz` case was not considered.